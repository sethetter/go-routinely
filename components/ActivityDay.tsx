import * as React from 'react'
import * as moment from 'moment'
import { filter } from 'lodash'

export interface ActivityDayProps {
  activityId: string
  day: Date
  logsForDay: ActivityLog[]
  createLog: (params: Partial<ActivityLog>) => void
}

function logCountForActivityOnDay(
  day: Date,
  activityId: string,
  logs: ActivityLog[]
): number {
  return filter(logs, l => {
    const onSameDay = moment(l.completedAt).isSame(day, 'day')
    return l.activityId === activityId && onSameDay
  }).length
}

const ActivityDay = ({ day, activityId, logsForDay, createLog }: ActivityDayProps) => {
  const stars = []
  const logCount = logCountForActivityOnDay(day, activityId, logsForDay)

  for (let i = 0; i < logCount; i++) {
    stars.push(<i key={i} className="routinely--LogStar fa fa-star" />)
  }

  const createLogForDay = () => {
    createLog({
      activityId,
      completedAt: moment(day).startOf('day').add(12, 'hours').toDate()
    })
  }

  const className = `activity-day routinely--ActivityDayCol-${moment(day).weekday()}`

  return <td onClick={createLogForDay} className={className}>{stars}</td>
}

export default ActivityDay
