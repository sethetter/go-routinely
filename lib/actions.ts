import { Action } from 'redux'
import { Dispatch } from 'react-redux'

// TODO: replace this eventually
import * as api from './fake-api'

export type RoutinelyAction
  = ReceiveUserAction
  | RequestActivitiesAction
  | ReceiveActivitiesAction
  | RequestActivityLogsAction
  | ReceiveActivityLogsAction

export const enum Actions {
  ReceiveUser = 'Routinely/ReceiveUser',
  RequestActivities = 'Routinely/RequestActivities',
  ReceiveActivities = 'Routinely/ReceiveActivities',
  RequestActivityLogs = 'Routinely/RequestActivityLogs',
  ReceiveActivityLogs = 'Routinely/ReceiveActivityLogs'
}

export interface ReceiveUserAction extends Action {
  type: Actions.ReceiveUser
  user: UserData
}

export interface RequestActivityLogsAction extends Action {
  type: Actions.RequestActivityLogs
}

export interface ReceiveActivityLogsAction extends Action {
  type: Actions.ReceiveActivityLogs
  activityLogs: ActivityLog[]
}

export interface RequestActivitiesAction extends Action {
  type: Actions.RequestActivities
}

export interface ReceiveActivitiesAction extends Action {
  type: Actions.ReceiveActivities
  activities: Activity[]
}

/**
 * Action Creators
 */
export const receiveUser = (user: UserData): ReceiveUserAction => ({
  type: Actions.ReceiveUser,
  user
})

export const requestActivities = (): RequestActivitiesAction => ({
  type: Actions.RequestActivities
})

export const receiveActivities = (
  activities: Activity[]
): ReceiveActivitiesAction => ({
  type: Actions.ReceiveActivities,
  activities
})

export const requestActivityLogs = (): RequestActivityLogsAction => ({
  type: Actions.RequestActivityLogs
})

export const receiveActivityLogs = (
  activityLogs: ActivityLog[]
): ReceiveActivityLogsAction => ({
  type: Actions.ReceiveActivityLogs,
  activityLogs
})

export const loadActivities = () => (dispatch: Dispatch<ActivityLog[]>) => {
  dispatch(requestActivities())

  const activities = api.getActivities()
  dispatch(receiveActivities(activities))
}

export const loadActivityLogs = () => (dispatch: Dispatch<ActivityLog[]>) => {
  dispatch(requestActivityLogs())

  const activityLogs = api.getActivityLogs()
  dispatch(receiveActivityLogs(activityLogs))
}

export const createActivityLog = (
  activity: Activity,
  timestamp: Date
) => (
  dispatch: Dispatch<ActivityLog>
) => {
  const activityLogs = api.getActivityLogs()

  const newLog: ActivityLog = {
    id: activityLogs.length,
    name: activity.name,
    value: activity.value,
    activityId: activity.id,
    occurredAt: timestamp
  }

  api.saveActivityLog(newLog)

  dispatch(loadActivityLogs())
}
