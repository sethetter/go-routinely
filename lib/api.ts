import * as moment from 'moment'
import * as URI from 'urijs'
import * as fetch from 'isomorphic-fetch'

const defaultFetchOpts: RequestInit = {
  headers: { 'Content-Type': 'application/json '},
  credentials: 'same-origin',
}

const apiUrl = (isServer: boolean, path: string): string =>
  isServer ? `http://localhost:3000${path}` : path

export async function getActivities (
  isServer: boolean = false,
  headers: { [key: string]: string } = {},
): Promise<Activity[]> {
  const uri = new URI(apiUrl(isServer, '/api/activities'))

  return fetch(uri.valueOf(), { ...defaultFetchOpts, headers }).then(r => r.json())
}

export async function getLogsForWeek (
  date: Date,
  isServer: boolean = false,
  headers: { [key: string]: string } = {},
): Promise<ActivityLog[]> {
  const uri = new URI(apiUrl(isServer, '/api/logs'))
  uri.addQuery('week', moment(date).toISOString())

  return fetch(uri.valueOf(), { ...defaultFetchOpts, headers }).then(r => r.json())
}

export async function getUserData (
  isServer: boolean = false,
  headers: { [key: string]: string } = {},
): Promise<UserData | undefined> {
  const uri = new URI(apiUrl(isServer, '/api/user/me'))
  return fetch(uri.valueOf(), { ...defaultFetchOpts, headers }).then(r => r.json())
}

export async function createActivity (
  params: Partial<Activity>,
  isServer: boolean = false,
): Promise<Activity> {
  const uri = new URI(apiUrl(isServer, '/api/activities'))
  return fetch(uri.valueOf(), Object.assign({}, defaultFetchOpts, {
    body: JSON.stringify(params),
    method: 'POST',
  })).then(r => r.json())
}

export async function createLog (
  params: Partial<ActivityLog>,
  isServer: boolean = false,
): Promise<ActivityLog> {
  const uri = new URI(apiUrl(isServer, '/api/logs'))
  return fetch(uri.valueOf(), Object.assign({}, defaultFetchOpts, {
    body: JSON.stringify(params),
    method: 'POST',
  })).then(r => r.json())
}

export async function getPoints (
  isServer: boolean = false,
  headers: { [key: string]: string } = {},
): Promise<number> {
  const uri = new URI(apiUrl(isServer, '/api/points'))

  const { points } = await fetch(
    uri.valueOf(),
    Object.assign({}, defaultFetchOpts, { headers })
  ).then(r => r.json()) as any

  return points
}
