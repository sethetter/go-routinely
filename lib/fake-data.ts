import * as moment from 'moment'

interface AppData {
  startOfWeek: Date
  activities: Activity[]
  activityLogs: ActivityLog[]
  weekDays: Date[]
}

const startOfWeek = new Date('03/25/2018')

const WEEKDAYS = [
  moment(startOfWeek).day('Sunday').startOf('day').toDate(),
  moment(startOfWeek).day('Monday').startOf('day').toDate(),
  moment(startOfWeek).day('Tuesday').startOf('day').toDate(),
  moment(startOfWeek).day('Wednesday').startOf('day').toDate(),
  moment(startOfWeek).day('Thursday').startOf('day').toDate(),
  moment(startOfWeek).day('Friday').startOf('day').toDate(),
  moment(startOfWeek).day('Saturday').startOf('day').toDate(),
]

const DATA: AppData = {
  startOfWeek,
  activities: [
    { id: 1, name: 'Brush teeth', value: 1 },
    { id: 2, name: 'Taekwondo', value: 2 },
    { id: 3, name: 'Clean room', value: 3 },
  ],
  activityLogs: [
    { id: 1, name: 'Brush teeth', value: 1, activityId: 1, occurredAt: WEEKDAYS[0] },
    { id: 2, name: 'Brush teeth', value: 1, activityId: 1, occurredAt: WEEKDAYS[0] },
    { id: 3, name: 'Taekwondo', value: 2, activityId: 2, occurredAt: WEEKDAYS[3] },
    { id: 4, name: 'Clean room', value: 3, activityId: 3, occurredAt: WEEKDAYS[2] },
  ],
  weekDays: WEEKDAYS
}

export const getFakeData = (): AppData => Object.assign({}, DATA)