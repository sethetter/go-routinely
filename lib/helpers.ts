import * as moment from 'moment'

export function startOfWeekFromDate (date: Date): Date {
  return moment(date).day('Sunday').startOf('day').toDate()
}

export function weekDaysFromDate (date: Date): Date[] {
  // Subtract one to start so loop can always add one
  const startOfWeek = moment(date).day('Sunday').subtract(1, 'day').startOf('day')
  const weekDays = []

  for (let i = 0; i < 7; i++) {
    weekDays.push(startOfWeek.add(1, 'day').toDate())
  }

  return weekDays
}
