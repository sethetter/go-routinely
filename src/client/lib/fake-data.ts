import moment from 'moment'

interface AppData {
  user: UserData
  startOfWeek: Date
  activities: Activity[]
  activityLogs: ActivityLog[]
  weekDays: Date[]
}

const startOfWeek = new Date('03/25/2018')

const WEEKDAYS = [
  moment(startOfWeek).day('Sunday').startOf('day').toDate(),
  moment(startOfWeek).day('Monday').startOf('day').toDate(),
  moment(startOfWeek).day('Tuesday').startOf('day').toDate(),
  moment(startOfWeek).day('Wednesday').startOf('day').toDate(),
  moment(startOfWeek).day('Thursday').startOf('day').toDate(),
  moment(startOfWeek).day('Friday').startOf('day').toDate(),
  moment(startOfWeek).day('Saturday').startOf('day').toDate(),
]

const DATA: AppData = {
  startOfWeek,
  user: { id: 'aosidfjoasidfj' },
  activities: [
    { _id: '1', name: 'Brush teeth', value: 1 },
    { _id: '2', name: 'Taekwondo', value: 2 },
    { _id: '3', name: 'Clean room', value: 3 },
  ],
  activityLogs: [
    { _id: '1', name: 'Brush teeth', value: 1, activityId: '1', occurredAt: WEEKDAYS[0] },
    { _id: '2', name: 'Brush teeth', value: 1, activityId: '1', occurredAt: WEEKDAYS[0] },
    { _id: '3', name: 'Taekwondo', value: 2, activityId: '2', occurredAt: WEEKDAYS[3] },
    { _id: '4', name: 'Clean room', value: 3, activityId: '3', occurredAt: WEEKDAYS[2] },
  ],
  weekDays: WEEKDAYS
}

export const getFakeData = (): AppData => Object.assign({}, DATA)
