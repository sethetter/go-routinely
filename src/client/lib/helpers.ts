import * as moment from 'moment'

export function startOfWeekFromDate (date: Date): Date {
  return moment(middleOfDay(date)).day('Sunday').toDate()
}

export function weekDaysFromDate (date: Date): Date[] {
  // Subtract one to start so loop can always add one
  const startOfWeek = moment(middleOfDay(date)).subtract(1, 'day')
  const weekDays = []

  for (let i = 0; i < 7; i++) {
    weekDays.push(startOfWeek.add(1, 'day').toDate())
  }

  return weekDays
}

export function middleOfDay (date: Date): Date {
  return moment.utc(date).startOf('day').add(12, 'hours').toDate()
}
