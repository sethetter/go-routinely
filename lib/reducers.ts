import * as moment from 'moment'
import { toEntityStore } from './helpers'

import {
  RoutinelyAction,
  Actions
} from './actions'

export const initialState: RootState = {
  activities: {
    allIds: [],
    byId: {}
  },
  logs: {
    allIds: [],
    byId: {}
  },
  isFetching: false,
  startOfWeek: moment().startOf('week').toDate(),
  lastUpdated: moment().toDate()
}

// TODO: Rewrite this things!
function rootReducer (state: RootState = initialState, action: RoutinelyAction) {
  switch (action.type) {

    case Actions.RequestActivities:
      return Object.assign({}, {
        ...state,
        isFetching: true
      })

    case Actions.ReceiveActivities:
      return Object.assign({}, {
        ...state,
        isFetching: false,
        activities: toEntityStore(action.activities)
      })

    case Actions.RequestActivityLogs:
      return Object.assign({}, {
        ...state,
        isFetching: true
      })

    case Actions.ReceiveActivityLogs:
      return Object.assign({}, {
        ...state,
        isFetching: false,
        logs: toEntityStore(action.activityLogs)
      })

    default:
      return state
  }
}

export default rootReducer
