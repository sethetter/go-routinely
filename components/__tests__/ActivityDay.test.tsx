import * as React from 'react'
import * as moment from 'moment'

import { shallow } from 'enzyme'

import { middleOfDay } from '../../lib/helpers'
import { getFakeData } from '../../lib/fake-data'

import ActivityDay, { ActivityDayProps } from '../ActivityDay'

const DATA = getFakeData()

const props: ActivityDayProps = {
  activityId: '1',
  day: new Date('2018-03-26'),
  logsForDay: DATA.activityLogs,
  createLog: jest.fn()
}

describe('<ActivityDay />', () => {
  it('renders without crashing', () => {
    shallow(<ActivityDay {...props} />)
  })

  it('shows a star for each log in logCount', () => {
    const day = moment(DATA.startOfWeek)
      .day('Sunday')
      .startOf('day')
      .toDate()

    const component = shallow(<ActivityDay {...props} day={day} />)

    expect(component).toMatchSnapshot()

    const stars = component.find('.routinely--LogStar')

    const expectedStarCount = DATA.activityLogs.filter(l => {
      const onSameDay = moment(l.completedAt).isSame(day, 'day')
      return l.activityId === props.activityId && onSameDay
    }).length

    expect(stars).toHaveLength(expectedStarCount)
  })

  it('creates a log for the specified day and activity', () => {
    const createLog = jest.fn()
    const component = shallow(<ActivityDay {...props} createLog={createLog} />)

    component.find('.activity-day').simulate('click')

    expect(createLog).toHaveBeenCalledWith({
      activityId: props.activityId,
      completedAt: middleOfDay(props.day),
    })
  })
})
