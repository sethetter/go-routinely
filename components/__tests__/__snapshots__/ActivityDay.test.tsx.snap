// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows a star for each log in logCount 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {},
        "children": Array [
          Object {
            "attribs": Object {},
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "routinely--ActivityDayCol-0",
                },
                "children": Array [
                  Object {
                    "attribs": Object {
                      "class": "routinely--LogStar fa fa-star",
                    },
                    "children": Array [],
                    "name": "i",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": Object {
                      "attribs": Object {
                        "class": "routinely--LogStar fa fa-star",
                      },
                      "children": Array [],
                      "name": "i",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": null,
                      "parent": [Circular],
                      "prev": [Circular],
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  Object {
                    "attribs": Object {
                      "class": "routinely--LogStar fa fa-star",
                    },
                    "children": Array [],
                    "name": "i",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": Object {
                      "attribs": Object {
                        "class": "routinely--LogStar fa fa-star",
                      },
                      "children": Array [],
                      "name": "i",
                      "namespace": "http://www.w3.org/1999/xhtml",
                      "next": [Circular],
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "class": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "class": undefined,
                      },
                    },
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                ],
                "name": "td",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "tr",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
        ],
        "name": "tbody",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {},
        "x-attribsPrefix": Object {},
      },
    ],
    "name": "table",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
